test/model.test.js

```diff
@@ -8859,6 +8859,35 @@ describe('Model', function() {
       assert.equal(error.errors['subdoc.num'].name, 'CastError');
       assert.equal(error.errors['docArr.0.num'].name, 'CastError');
     });
+    it('should not throw an error if `ignoreCastErrors` is set (gh-12156)', function() {
+      const Test = db.model('Test', mongoose.Schema({
+        _id: false,
+        num: Number,
+        nested: {
+          num: Number
+        },
+        subdoc: {
+          type: mongoose.Schema({
+            _id: false,
+            num: Number
+          }),
+          default: () => ({})
+        },
+        docArr: [{
+          _id: false,
+          num: Number
+        }]
+      }));
+
+      const obj = {
+        num: 'not a number',
+        nested: { num: '2' },
+        subdoc: { num: '3' },
+        docArr: [{ num: '4' }]
+      };
+      const ret = Test.castObject(obj, { ignoreCastErrors: true });
+      assert.ok(ret);
```

Comment:
Would be nice if we assert that castObject returns the castable paths, and ignores the invalid ones.
```suggestion
      const obj = {
        num: 'not a number',
        nested: { num: '2' },
        subdoc: 'not a subdoc',
        docArr: [{ num: '4' }]
      };
      const ret = Test.castObject(obj, { ignoreCastErrors: true });
      assert.deepStrictEqual(ret, { nested: { num: 2 }, docArr: [{ num: 4 }] });
```